src/providers/idp/idp_init.c

```diff
@@ -0,0 +1,303 @@
+                         struct be_ctx *be_ctx,
+                         void *module_data,
+                         struct dp_method *dp_methods)
+{
+    struct idp_init_ctx *init_ctx;
+    struct idp_id_ctx *id_ctx;
+    errno_t ret;
+    enum idmap_error_code err;
+    struct sss_idmap_range id_range;
+
+    init_ctx = talloc_get_type(module_data, struct idp_init_ctx);
+
+    id_ctx = talloc_zero(init_ctx, struct idp_id_ctx);
+    if (id_ctx == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Failed to allocate memory for id context.\n");
+        return ENOMEM;
+    }
+
+    id_ctx->be_ctx = be_ctx;
+    id_ctx->init_ctx = init_ctx;
+    id_ctx->idp_options = init_ctx->opts;
+
+    id_ctx->idp_type = init_ctx->idp_type;
+    id_ctx->client_id = init_ctx->client_id;
+    id_ctx->client_secret = init_ctx->client_secret;
+    id_ctx->token_endpoint = init_ctx->token_endpoint;
+    id_ctx->scope = init_ctx->scope;
+
+    err = sss_idmap_init(sss_idmap_talloc, init_ctx, sss_idmap_talloc_free,
+                         &id_ctx->idmap_ctx);
+    if (err != IDMAP_SUCCESS) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Failed in initialize id-mapping: [%s].\n",
+                                   idmap_error_string(err));
+        ret = EINVAL;
+        goto done;
+    }
+
+    /* TODO: The range_id (2nd parameter) should be configurable */
+    err = sss_idmap_calculate_range(id_ctx->idmap_ctx, id_ctx->token_endpoint,
+                                    NULL, &id_range);
+    if (err != IDMAP_SUCCESS) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Failed to calculate id range for [%s]: [%s].\n",
+              id_ctx->token_endpoint, idmap_error_string(err));
+        ret = EINVAL;
+        goto done;
+    }
+
+    err = sss_idmap_add_gen_domain_ex(id_ctx->idmap_ctx, be_ctx->domain->name,
+                                      id_ctx->token_endpoint, &id_range,
+                                      NULL, NULL, NULL, NULL, 0, false);
+    if (err != IDMAP_SUCCESS) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Failed to add id-mapping domain [%s]: [%s].\n",
+              be_ctx->domain->name, idmap_error_string(err));
+        ret = EINVAL;
+        goto done;
+    }
+
+
+    dp_set_method(dp_methods, DPM_ACCOUNT_HANDLER,
+                  idp_account_info_handler_send, idp_account_info_handler_recv, id_ctx,
+                  struct idp_id_ctx, struct dp_id_data, struct dp_reply_std);
+
+    dp_set_method(dp_methods, DPM_CHECK_ONLINE,
+                  idp_online_check_handler_send, idp_online_check_handler_recv, id_ctx,
+                  struct idp_id_ctx, void, struct dp_reply_std);
+
+    dp_set_method(dp_methods, DPM_ACCT_DOMAIN_HANDLER,
+                  default_account_domain_send, default_account_domain_recv, NULL,
+                  void, struct dp_get_acct_domain_data, struct dp_reply_std);
+
+    ret = EOK;
+
+done:
+    if (ret != EOK) {
+        talloc_free(id_ctx);
+    }
+
+    return ret;
```

Comment:
Isn't it required to store `init_ctx->id_ctx = id_ctx`?

Reply:
Currently the IdP provider does not need it, but to be in line with the other providers I added it.